File: vespajlib/src/main/java/com/yahoo/slime/BinaryEncoder.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.slime;

import static com.yahoo.slime.BinaryFormat.encode_double;
import static com.yahoo.slime.BinaryFormat.encode_type_and_meta;
import static com.yahoo.slime.BinaryFormat.encode_zigzag;

final class BinaryEncoder implements ArrayTraverser, ObjectSymbolTraverser {

    private final BufferedOutput out;

    BinaryEncoder() {
        this(new BufferedOutput());
    }
    BinaryEncoder(BufferedOutput output) {
        out = output;
    }

    BufferedOutput encode(Slime slime) {
        out.reset();
        encodeSymbolTable(slime);
        encodeValue(slime.get());
        return out;
    }


    void encode_cmpr_int(int value) {
        byte next = (byte)(value & 0x7f);
        value >>>= 7; // unsigned shift
        while (value != 0) {
            next |= (byte)0x80;
            out.put(next);
            next = (byte)(value & 0x7f);
            value >>>= 7;
        }
        out.put(next);
    }

    void write_type_and_size(int type, int size) {
        if (size <= 30) {
            out.put(encode_type_and_meta(type, size + 1));
        } else {
            out.put(encode_type_and_meta(type, 0));
            encode_cmpr_int(size);
        }
    }

    void write_type_and_bytes_le(int type, long bits) {
        int pos = out.position();
        byte val = 0;
        out.put(val);
        while (bits != 0) {
            val = (byte)(bits & 0xff);
            bits >>>= 8;
            out.put(val);
        }
        val = encode_type_and_meta(type, out.position() - pos - 1);
        out.absolutePut(pos, val);
    }

    void write_type_and_bytes_be(int type, long bits) {
        int pos = out.position();
        byte val = 0;
        out.put(val);
        while (bits != 0) {
            val = (byte)(bits >> 56);
            bits <<= 8;
            out.put(val);
        }
        val = encode_type_and_meta(type, out.position() - pos - 1);
        out.absolutePut(pos, val);
    }

    void encodeNIX() {
        out.put(Type.NIX.ID);
    }

    void encodeBOOL(boolean value) {
        out.put(encode_type_and_meta(Type.BOOL.ID, value ? 1 : 0));
    }

    void encodeLONG(long value) {
        write_type_and_bytes_le(Type.LONG.ID, encode_zigzag(value));
    }

    void encodeDOUBLE(double value) {
        write_type_and_bytes_be(Type.DOUBLE.ID, encode_double(value));
    }

    void encodeSTRING(byte[] value) {
        write_type_and_size(Type.STRING.ID, value.length);
        out.put(value);
    }

    void encodeDATA(byte[] value) {
        write_type_and_size(Type.DATA.ID, value.length);
        out.put(value);
    }

    void encodeARRAY(Inspector inspector) {
        write_type_and_size(Type.ARRAY.ID, inspector.children());
        ArrayTraverser at = this;
        inspector.traverse(at);
    }

    void encodeOBJECT(Inspector inspector) {
        write_type_and_size(Type.OBJECT.ID, inspector.children());
        ObjectSymbolTraverser ot = this;
        inspector.traverse(ot);
    }

    void encodeValue(Inspector inspector) {
        switch(inspector.type()) {
        case NIX:    encodeNIX();                        return;
        case BOOL:   encodeBOOL(inspector.asBool());     return;
        case LONG:   encodeLONG(inspector.asLong());     return;
        case DOUBLE: encodeDOUBLE(inspector.asDouble()); return;
        case STRING: encodeSTRING(inspector.asUtf8());   return;
        case DATA:   encodeDATA(inspector.asData());     return;
        case ARRAY:  encodeARRAY(inspector);             return;
        case OBJECT: encodeOBJECT(inspector);            return;
        }
        assert false : "Should not be reached";
    }

    void encodeSymbolTable(Slime slime) {
        int numSymbols = slime.symbols();
        encode_cmpr_int(numSymbols);
        for (int i = 0 ; i < numSymbols; ++i) {
            String name = slime.inspect(i);
            byte[] bytes = Utf8Codec.encode(name);
            encode_cmpr_int(bytes.length);
            out.put(bytes);
        }
    }

    public void entry(int idx, Inspector inspector) {
        encodeValue(inspector);
    }

    public void field(int symbol, Inspector inspector) {
        encode_cmpr_int(symbol);
        encodeValue(inspector);
    }

}

File: client/go/internal/slime/binary_encoder.go
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
// Author: arnej

package slime

type binaryEncoder struct {
	buf     []byte
	symbols map[string]uint32
}

func newBinaryEncoder() binaryEncoder {
	return binaryEncoder{
		buf:     make([]byte, 0, 128),
		symbols: make(map[string]uint32),
	}
}

func (out *binaryEncoder) pos() int {
	return len(out.buf)
}

func (out *binaryEncoder) put(b byte) {
	out.buf = append(out.buf, b)
}

func (out *binaryEncoder) putAt(pos int, b byte) {
	out.buf[pos] = b
}

func (out *binaryEncoder) putBytes(bytes []byte) {
	out.buf = append(out.buf, bytes...)
}

func (out *binaryEncoder) putString(value string) {
	out.buf = append(out.buf, value...)
}

func (out *binaryEncoder) encode_cmpr_uint(value uint32) {
	var next byte = byte(value & 0x7f)
	value = value >> 7
	for value != 0 {
		next = next | 0x80
		out.put(next)
		next = byte(value & 0x7f)
		value = value >> 7
	}
	out.put(next)
}

func (out *binaryEncoder) write_type_and_size(ty Type, size uint32) {
	if size <= 30 {
		out.put(encode_type_and_meta(ty, size+1))
	} else {
		out.put(encode_type_and_meta(ty, 0))
		out.encode_cmpr_uint(size)
	}
}

func (out *binaryEncoder) write_type_and_bytes_le(ty Type, bits uint64) {
	pos := out.pos()
	var val byte = 0
	out.put(val)
	var cnt uint32 = 0
	for bits != 0 {
		val = byte(bits & 0xff)
		bits = bits >> 8
		out.put(val)
		cnt++
	}
	val = encode_type_and_meta(ty, cnt)
	out.putAt(pos, val)
}

func (out *binaryEncoder) write_type_and_bytes_be(ty Type, bits uint64) {
	pos := out.pos()
	var val byte = 0
	out.put(val)
	var cnt uint32 = 0
	for bits != 0 {
		val = byte(bits >> 56)
		bits = bits << 8
		out.put(val)
		cnt++
	}
	val = encode_type_and_meta(ty, cnt)
	out.putAt(pos, val)
}

func (out *binaryEncoder) encodeNIX() {
	out.put(byte(NIX))
}

func (out *binaryEncoder) encodeValue(inspector Inspector) {
	switch inspector.Type() {
	case NIX:
		out.encodeNIX()
	case BOOL:
		out.encodeBOOL(inspector.AsBool())
	case LONG:
		out.encodeLONG(inspector.AsLong())
	case DOUBLE:
		out.encodeDOUBLE(inspector.AsDouble())
	case STRING:
		out.encodeSTRING(inspector.AsString())
	case DATA:
		out.encodeDATA(inspector.AsData())
	case ARRAY:
		out.encodeARRAY(inspector)
	case OBJECT:
		out.encodeOBJECT(inspector)
	default:
		panic("Should not be reached")
	}
}

func (out *binaryEncoder) encodeBOOL(value bool) {
	var meta uint32
	if value {
		meta = 1
	} else {
		meta = 0
	}
	out.put(encode_type_and_meta(BOOL, meta))
}

func (out *binaryEncoder) encodeLONG(value int64) {
	out.write_type_and_bytes_le(LONG, encode_zigzag(value))
}

func (out *binaryEncoder) encodeDOUBLE(value float64) {
	out.write_type_and_bytes_be(DOUBLE, encode_double(value))
}

func (out *binaryEncoder) encodeSTRING(value string) {
	out.write_type_and_size(STRING, uint32(len(value)))
	out.putString(value)
}

func (out *binaryEncoder) encodeDATA(value []byte) {
	out.write_type_and_size(DATA, uint32(len(value)))
	out.putBytes(value)
}

func (out *binaryEncoder) Entry(idx int, value Inspector) {
	out.encodeValue(value)
}

func (out *binaryEncoder) encodeARRAY(inspector Inspector) {
	out.write_type_and_size(ARRAY, uint32(inspector.Entries()))
	inspector.TraverseArray(out)
}

func (out *binaryEncoder) findSymbolId(name string) uint32 {
	if value, ok := out.symbols[name]; ok {
		return value
	}
	value := uint32(len(out.symbols))
	out.symbols[name] = value
	return value
}

func (out *binaryEncoder) Field(name string, value Inspector) {
	symbol := out.findSymbolId(name)
	out.encode_cmpr_uint(symbol)
	out.encodeValue(value)
}

func (out *binaryEncoder) encodeOBJECT(inspector Inspector) {
	out.write_type_and_size(OBJECT, uint32(inspector.Fields()))
	inspector.TraverseObject(out)
}

func (out *binaryEncoder) prependSymbolTable() []byte {
	numSymbols := len(out.symbols)
	sorted := make([]string, numSymbols)
	for name, id := range out.symbols {
		sorted[id] = name
	}
	tmp := newBinaryEncoder()
	tmp.encode_cmpr_uint(uint32(numSymbols))
	for _, name := range sorted {
		sz := uint32(len(name))
		tmp.encode_cmpr_uint(sz)
		tmp.putString(name)
	}
	res := make([]byte, len(out.buf)+len(tmp.buf))
	split := len(tmp.buf)
	copy(res[:split], tmp.buf)
	copy(res[split:], out.buf)
	return res
}
